File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/controller/overlay/handler/InfoWindowHandler.kt
package dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.handler

import com.naver.maps.map.overlay.InfoWindow
import com.naver.maps.map.overlay.Overlay
import dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.OverlayHandler
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay.NInfoWindow
import io.flutter.plugin.common.MethodChannel

internal interface InfoWindowHandler : OverlayHandler {
    fun handleInfoWindow(
        infoWindow: Overlay, method: String, arg: Any?, result: MethodChannel.Result,
    ) = (infoWindow as InfoWindow).let { i ->
        when (method) {
            NInfoWindow.textName -> setText(i, arg!!)
            NInfoWindow.anchorName -> setAnchor(i, arg!!)
            NInfoWindow.alphaName -> setAlpha(i, arg!!)
            NInfoWindow.positionName -> setPosition(i, arg!!)
            NInfoWindow.offsetXName -> setOffsetX(i, arg!!)
            NInfoWindow.offsetYName -> setOffsetY(i, arg!!)
            NInfoWindow.closeName -> close(i)
            else -> result.notImplemented()
        }
    }

    fun setText(infoWindow: InfoWindow, rawText: Any)

    fun setAnchor(infoWindow: InfoWindow, rawNPoint: Any)

    fun setAlpha(infoWindow: InfoWindow, rawAlpha: Any)

    fun setPosition(infoWindow: InfoWindow, rawPosition: Any)

    fun setOffsetX(infoWindow: InfoWindow, rawOffsetXDp: Any)

    fun setOffsetY(infoWindow: InfoWindow, rawOffsetYDp: Any)

    fun close(infoWindow: InfoWindow)
}

File: ios/flutter_naver_map/Sources/flutter_naver_map/controller/overlay/handler/InfoWindowHandler.swift
import NMapsMap
import Flutter

internal protocol InfoWindowHandler: OverlayHandler {
        func setText(_ infoWindow: NMFInfoWindow, rawText: Any)

        func setAnchor(_ infoWindow: NMFInfoWindow, rawNPoint: Any)

        func setAlpha(_ infoWindow: NMFInfoWindow, rawAlpha: Any)

        func setPosition(_ infoWindow: NMFInfoWindow, rawPosition: Any)

        func setOffsetX(_ infoWindow: NMFInfoWindow, rawOffsetX: Any)

        func setOffsetY(_ infoWindow: NMFInfoWindow, rawOffsetY: Any)

        func close(_ infoWindow: NMFInfoWindow)
}

internal extension  InfoWindowHandler {
    func handleInfoWindow(infoWindow: NMFOverlay,
                             method: String,
                             args: Any?,
                             result: @escaping FlutterResult) {
        let infoWindow = infoWindow as! NMFInfoWindow
        switch method {
        case NInfoWindow.textName: setText(infoWindow, rawText: args!)
        case NInfoWindow.anchorName: setAnchor(infoWindow, rawNPoint: args!)
        case NInfoWindow.alphaName: setAlpha(infoWindow, rawAlpha: args!)
        case NInfoWindow.positionName: setPosition(infoWindow, rawPosition: args!)
        case NInfoWindow.offsetXName: setOffsetX(infoWindow, rawOffsetX: args!)
        case NInfoWindow.offsetYName: setOffsetY(infoWindow, rawOffsetY: args!)
        case NInfoWindow.closeName: close(infoWindow)
        default: result(FlutterMethodNotImplemented)
        }
    }
}
